File: core/src/test/java/com/xmartlabs/base/core/helpers/TestingTree.java
package com.xmartlabs.base.core.helpers;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.annimon.stream.Objects;
import com.annimon.stream.Stream;
import com.xmartlabs.base.core.observers.ObservableResult;

import java.util.ArrayList;
import java.util.List;

import lombok.Getter;
import timber.log.Timber;

public class TestingTree extends Timber.Tree {
  public static final String DEFAULT_NOT_FOUND_TREE_NODE = "Tree node not found";
  private static final TestingTreeNode NODE_NOT_FOUND = TestingTreeNode.builder().tag(DEFAULT_NOT_FOUND_TREE_NODE)
      .build();
  private static final String INCORRECT_EXCEPTION_THROWN = "Incorrect exception thrown";

  @Getter
  private List<TestingTreeNode> loggingTree = new ArrayList<>();

  @Override
  protected void log(int priority, @Nullable String tag, @Nullable String message, @Nullable Throwable t) {
    TestingTreeNode loggingNode = TestingTreeNode.builder()
        .message(message)
        .priority(priority)
        .tag(tag)
        .throwable(t)
        .build();

    loggingTree.add(loggingNode);
  }

  @NonNull
  private TestingTreeNode getLogTreeNodeWithTag(@NonNull String tag) {
    return Stream.ofNullable(loggingTree)
        .filter(node -> Objects.equals(node.getTag(), tag))
        .findFirst()
        .orElse(NODE_NOT_FOUND);
  }

  @NonNull
  public TestingTreeNode getLogTreeNodeWithTag(@NonNull ObservableResult observableResult) {
    return getLogTreeNodeWithTag(observableResult.toString());
  }

  private long getLogTreeNodesCountWithTag(@NonNull String tag) {
    return Stream.ofNullable(loggingTree)
        .filter(node -> Objects.equals(node.getTag(), tag))
        .count();
  }

  public long getLogTreeNodesCountWithTag(@NonNull ObservableResult observableResultTag) {
    return getLogTreeNodesCountWithTag(observableResultTag.toString());
  }

  @NonNull
  private String getLogTreeExceptionDetailMessage(@NonNull final String detailMessage) {
    return Stream.ofNullable(loggingTree)
        .map(node -> node.getThrowable().getMessage())
        .filter(message -> Objects.equals(message, detailMessage))
        .findFirst()
        .orElse(INCORRECT_EXCEPTION_THROWN);
  }

  @NonNull
  public String getLogTreeExceptionDetailMessage(@NonNull final ObservableResult detailMessage) {
    return getLogTreeExceptionDetailMessage(detailMessage.toString());
  }

  public static void log(ObservableResult observableResult) {
    Timber.tag(observableResult.toString()).i(observableResult.toString());
  }
}

File: core/src/test/java/com/xmartlabs/bigbang/core/helpers/TestingTree.kt
package com.xmartlabs.bigbang.core.helpers

import com.xmartlabs.bigbang.core.observers.ObservableResult

import timber.log.Timber

class TestingTree : Timber.Tree() {
  companion object {
    val DEFAULT_NOT_FOUND_TREE_NODE = "Tree node not found"
    private val NODE_NOT_FOUND = TestingTreeNode(0, DEFAULT_NOT_FOUND_TREE_NODE)
    private val INCORRECT_EXCEPTION_THROWN = "Incorrect exception thrown"
    
    fun log(observableResult: ObservableResult) = Timber.tag(observableResult.toString()).i(observableResult.toString())
  }
  
  private val loggingTree = ArrayList<TestingTreeNode>()

  override fun log(priority: Int, tag: String?, message: String?, t: Throwable?) {
    loggingTree.add(TestingTreeNode(priority, tag, message, t))
  }

  private fun getLogTreeNodeWithTag(tag: String) =
      loggingTree.filter { it.tag == tag }.firstOrNull() ?: NODE_NOT_FOUND

  fun getLogTreeNodeWithTag(observableResult: ObservableResult) =
      getLogTreeNodeWithTag(observableResult.toString())

  private fun getLogTreeNodesCountWithTag(tag: String) = loggingTree.filter { it.tag == tag }.count()

  fun getLogTreeNodesCountWithTag(observableResultTag: ObservableResult) =
      getLogTreeNodesCountWithTag(observableResultTag.toString())

  private fun getLogTreeExceptionDetailMessage(detailMessage: String) = loggingTree
      .map { it.throwable?.message }
      .filterNotNull()
      .filter { it == detailMessage }
      .firstOrNull()
      ?: INCORRECT_EXCEPTION_THROWN

  fun getLogTreeExceptionDetailMessage(detailMessage: ObservableResult) =
      getLogTreeExceptionDetailMessage(detailMessage.toString())
}
